### PR TITLE
types: support converting json to float64

### LINKED ISSUE #6014: Can not support sum(...) on json data?
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
CREATE TABLE `t_logger_info` (
	 	  `id` varchar(64) NOT NULL,
	 	  `level` int NOT NULL,
	 	  `content` text NOT NULL,
	 	  `create_time` datetime NOT NULL,
	 	  `log_time` datetime NOT NULL,
          `tags` json DEFAULT NULL,
	 	  PRIMARY KEY (`id`)
	 	) ENGINE=InnoDB DEFAULT CHARSET=utf8;

and insert 1 millons rows, some like: 
5a966eb5ecf6a05fbc7e2d84 | 2 | test batch 0 | 2018-02-28 16:56:21 | 2018-02-28 16:56:21 | {"b": true, "i": 1, "app": "test-0", "os-name": 0}
-- | -- | -- | -- | -- | --
5a966eb5ecf6a05fbc7e2d85 | 2 | test batch 0 | 2018-02-28 16:56:21 | 2018-02-28 16:56:21 | {"b": false, "i": 2, "app": "test-0", "os-name": 0}
5a966eb5ecf6a05fbc7e2d86 | 2 | test batch 0 | 2018-02-28 16:56:21 | 2018-02-28 16:56:21 | {"b": true, "i": 3, "app": "test-0", "os-name": 0}
5a966eb5ecf6a05fbc7e2d87 | 2 | test batch 0 | 2018-02-28 16:56:21 | 2018-02-28 16:56:21 | {"b": false, "i": 4, "app": "test-0", "os-name": 0}
5a966eb5ecf6a05fbc7e2d88 | 2 | test batch 0 | 2018-02-28 16:56:21 | 2018-02-28 16:56:21 | {"b": true, "i": 5, "app": "test-0", "os-name": 0}
5a966eb5ecf6a05fbc7e2d89 | 2 | test batch 0 | 2018-02-28 16:56:21 | 2018-02-28 16:56:21 | {"b": false, "i": 6, "app": "test-0", "os-name": 0}
5a966eb5ecf6a05fbc7e2d8a | 2 | test batch 0 | 2018-02-28 16:56:21 | 2018-02-28 16:56:21 | {"b": true, "i": 7, "app": "test-0", "os-name": 0}
5a966eb5ecf6a05fbc7e2d8b | 2 | test batch 0 | 2018-02-28 16:56:21 | 2018-02-28 16:56:21 | {"b": false, "i": 8, "app": "test-0", "os-name": 0}
5a966eb5ecf6a05fbc7e2d8c | 2 | test batch 0 | 2018-02-28 16:56:21 | 2018-02-28 16:56:21 | {"b": true, "i": 9, "app": "test-0", "os-name": 0}
5a966eb5ecf6a05fbc7e2d8d | 2 | test batch 0 | 2018-02-28 16:56:21 | 2018-02-28 16:56:21 | {"b": false, "i": 10, "app": "test-0", "os-name": 0}

i use sum on json data:
SELECT  sum(tags->'$.i') FROM db_pangu.t_logger_info    

2. What did you expect to see?

in mysql:  give the result :  495000000
--
3. What did you see instead?

in tidb:  Error Code: 1105. cannot convert 0(type json.BinaryJSON) to float64

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
5.7.1-Tidb-v1.1.0-apha.1-85-gbe1eeac

### PR DESCRIPTION
**Pull request by yangwenmai**
fix #6014

### CODE REVIEW

**Comment by shenli** [comment]
@yangwenmai Thanks!
Is there a related issue about this bug? How do you find it?

**Comment by yangwenmai** [comment]
According to the error message:`ERROR 1105 (HY000): cannot convert 1(type json.BinaryJSON) to float64` and I find the code `https://github.com/pingcap/tidb/blob/master/types/datum.go#L1448`.
`func (d *Datum) ToFloat64(sc *stmtctx.StatementContext) (float64, error)` function does not handle KindMysqlJSON type.

**Comment by winkyao** [comment]
@yangwenmai Please add some related unit tests.

File: types/datum.go [hunk 1]
```diff
@@ -1444,6 +1444,9 @@ func (d *Datum) ToFloat64(sc *stmtctx.StatementContext) (float64, error) {
 	case KindBinaryLiteral, KindMysqlBit:
 		val, err := d.GetBinaryLiteral().ToInt()
 		return float64(val), err
+	case KindMysqlJSON:
+		f, err := ConvertJSONToFloat(sc, d.GetMysqlJSON())
+		return f, err
```
Comment (zz-jason) [inline types/datum.go]: We should return `f, errors.Trace(err)` to track the whole error stack. You can modify lIne 1446 together.

**Comment by yangwenmai** [comment]
@zz-jason PTAL

File: types/datum.go [hunk 2]
```diff
@@ -1513,6 +1516,10 @@ func (d *Datum) ToMysqlJSON() (j json.BinaryJSON, err error) {
 		in = d.GetFloat64()
 	case KindMysqlDecimal:
 		in, err = d.GetMysqlDecimal().ToFloat64()
+		if err != nil {
```
Comment (winkyao) [inline types/datum.go]: Not need to return here. It will return in line 1537.

File: types/datum.go [hunk 3]
```diff
@@ -1586,9 +1593,12 @@ func CoerceDatum(sc *stmtctx.StatementContext, a, b Datum) (x, y Datum, err erro
 		case KindUint64:
 			y.SetFloat64(float64(y.GetUint64()))
 		case KindBinaryLiteral, KindMysqlBit:
-			val, err1 := y.GetBinaryLiteral().ToInt()
-			err = err1
-			y.SetFloat64(float64(val))
+			var fval uint64
+			fval, err = y.GetBinaryLiteral().ToInt()
+			if err != nil {
```
Comment (winkyao) [inline types/datum.go]: ditto.

Comment (yangwenmai) [reply types/datum.go]: It not need return in line 1610?

File: types/datum.go (hunk 2 above)
Comment (yangwenmai) [reply types/datum.go]: It not need to return in line 1533?

Comment (winkyao) [reply types/datum.go]: you can add 
```
if err != nil {
   err = errros.Trace(err)
   return
}
```

at the end of the function.

Comment (winkyao) [reply types/datum.go]: and line 1533 can be removed.

File: types/datum.go (hunk 3 above)
Comment (winkyao) [reply types/datum.go]: Got it.

**Pull request merged**